### PR TITLE
New Machine Builder API

### PR DESCRIPTION
**Pull request by xrstf**
**What this PR does / why we need it**:
The old way we assembled ProviderSpec for MachineDeployments was to rely on API types (like apiv1.NodeSpec). These types were removed and KKP is being refactored to not rely on API types anymore.

To that end, this PR consolidates the remaining places where we assemble Machines and provide a new Go API to make creating machines much easier (hopefully).

### `pkg/machine/`

This is where the common helper functions to deal with the machine-controller types live. Here you can

* translate the KKP provider name into the MC provider name using `MachineControllerProviderName()`,
* do the opposite via `KubermaticProviderType()`,
* determine the OS name based on the spec struct, using `OperatingSystemFromSpec()`,
* do the same for provider specs using `ProviderTypeFromSpec()`,
* etc.

This package also contains the `MachineBuilder`. This new struct defines a fluent API to assemble everything needed to ultimately create a MachineSpec. The builder combines OS, OS spec, networking config, provider spec etc. and can spit out ProviderSpec, ProviderConfig and CloudProviderSpecs (yes, the naming here is super confusing and not 100% consistent).

The `MachineBuilder` is **not necessary** to build machines, it only makes building them more convenient. Ultimately the builder "just" calls the various functions like `CompleteCloudProviderSpec()` and the functions mentioned above to try to intelligently combine everything.

### `pkg/machine/operatingsystem/`

This package defines a very basic set of builders for each OS. They are _very_ basic and only provide a super thin fluent API for assembling OS specs. Of most interest is the `DefaultSpec()` func, which will give you a default spec for the given OS and provider.

All of these helpers are optional and the `MachineBuilder` works with simple OS specs, so it can be used without the OS builders.

```go
builder := machine.NewBuilder().
   WithOperatingSystemSpec(k8c.io/osm/ubuntu.Config{DistUpgradeOnBoot: true})

// is identical to

builder := machine.NewBuilder().
   WithOperatingSystemSpec(os.NewUbuntuSpecBuilder("").WithDistUpgradeOnBoot(true).Build())
```

Again, a very thin helper that does not offer much, but also costs very little to maintain.

### `pkg/machine/provider/`

This package contains code for each provider which will combine the KKP Cluster object, the KKP datacenter and (optionally) an existing provider spec into a "filled-in" spec. So for example the `CompleteAWSCloudProviderSpec()` will apply the configured AWS region from the datacenter to the providerSpec (unless it is alreay given).

In addition to a `Complete...Spec()` function, each provider also comes with a very thin fluent builder API, which (compared to the operating system ones) is more useful, as it helps avoid dealing with the providerconfig.ConfigVarString thingies. But as with OS, the builders are completely optional.

```go
builder := machine.NewBuilder().
   WithCloudProviderSpec(k8c.io/mc/aws.RawConfig{
      VpcID:        providertypes.ConfigVarString{Value: "foo-vpc"},
      InstanceType: providertypes.ConfigVarString{Value: "t3.small"},
   })

// is identical to

builder := machine.NewBuilder().
   WithCloudProviderSpec(provider.NewAWSConfig().
      WithVPC("foo-vpc").
      WithInstanceType("t3.small").
      Build())
```

The functions are meant to fill in missing values. This is useful for the initial-machinedeployment controller, which takes the initial MD spec from an annotation and then applies it to the usercluster. But before applying it, it runs it once through the `Complete` function for the chosen provider, which will default values are applied properly.

For each provider I also defined a bunch of (mostly identical) unit tests to ensure the basics work as expected. This is why this PR adds so much new code. Be warned though, the tests make heavy use of generics and it might not be easy to immediately follow what's going on behind the scenes. But the upside was that tests are much easier to write and I hope this encourages us to write more when needed.

### Closing Remarks

Only 2 places in this repository even make use of this code: the conformance-tester's scenarios and the e2e test jig. But I expect that maybe the dashboard will start to import and use it, but this will require some refactoring and might not even be feasable.

**What type of PR is this?**
/kind cleanup

**Does this PR introduce a user-facing change? Then add your Release Note here**:
```release-note
Instead of an `apiv1.NodeDeployment`, a `clusterv1alpha1.MachineDeployment` must be stored in the `kubermatic.io/initial-machinedeployment-request` annotation on new clusters.
```

**Documentation**:
```documentation
NONE
```

### CODE REVIEW

File: hack/ci/run-conformance-tests.sh [hunk 1]
```diff
@@ -52,7 +52,7 @@ elif [[ $provider == "packet" ]]; then
   EXTRA_ARGS="-packet-api-key=${PACKET_API_KEY}
     -packet-project-id=${PACKET_PROJECT_ID}
     -packet-kkp-datacenter=packet-am"
-elif [[ $provider == "gcp" ]] || [[ $provider == "gce" ]]; then
+elif [[ $provider == "gcp" ]] || [[ $provider == "gcp" ]]; then
```
Comment (pkprzekwas) [inline hack/ci/run-conformance-tests.sh]: Looks like some leftover of replace function
```suggestion
elif [[ $provider == "gcp" ]]; then
```

File: hack/run-conformance-tests.sh [hunk 2]
```diff
@@ -100,7 +100,7 @@ digitalocean)
     -digitalocean-kkp-datacenter=do-ams3"
   ;;
 
-gce)
+gcp)
```
Comment (pkprzekwas) [inline hack/run-conformance-tests.sh:103]: This case is duplicated.

File: pkg/machine/builder.go [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+		return nil, nil
+	}
+
+	datacenterName := b.datacenterName
+	if datacenterName == "" && b.cluster != nil {
+		if b.cluster.Spec.Cloud.DatacenterName != b.datacenterName {
+			return nil, fmt.Errorf("explicit datacenter %q does not match cluster's datacenter %q", b.datacenterName, b.cluster.Spec.Cloud.DatacenterName)
+		}
+		datacenterName = b.cluster.Spec.Cloud.DatacenterName
+	}
```
Comment (pkprzekwas) [inline pkg/machine/builder.go]: I'm not entirely sure if I understand this part of code. I would expect a different condition in the external `if` statement. Please double check me on that.
```suggestion
	datacenterName := b.datacenterName
	if datacenterName != "" && b.cluster != nil {
		if b.cluster.Spec.Cloud.DatacenterName != b.datacenterName {
			return nil, fmt.Errorf("explicit datacenter %q does not match cluster's datacenter %q", b.datacenterName, b.cluster.Spec.Cloud.DatacenterName)
		}
		datacenterName = b.cluster.Spec.Cloud.DatacenterName
	}
```

**Review by pkprzekwas — CHANGES_REQUESTED** [review]

File: pkg/machine/builder.go (hunk 3 above)
Comment (xrstf) [reply pkg/machine/builder.go]: Damn you're good. Thaat was indeed broken. I added some unit tests and fixed the function, should now be a bit easier to read.

**Pull request merged**